File: 11-container-with-most-water/11-container-with-most-water.rb
# @param {Integer[]} height
# @return {Integer}
def max_area(height)
    l, r, ans = 0, height.length - 1, 0
    while l < r
        ans = [ans, (r-l) * [height[r], height[l]].min].max
        curr_l, curr_r = height[l], height[r]
        if height[l] <= height[r]
            while l < r && height[l] <= curr_l
                l += 1
            end
        else
            while l < r && height[r] <= curr_r
                r -= 1
            end
        end
    end
    return ans
end

File: 11-container-with-most-water/11-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r, ans = 0, len(height)-1, 0
        while l < r:
            ans = max(ans, (r-l) * min(height[r], height[l]))
            if height[l] <= height[r]:
                l += 1
            else:
                r -= 1
        return ans